### PR TITLE
Remove brush shape from label UI

### PR DESCRIPTION
**Pull request by ppwadhwa**
# Description

This issue: https://github.com/napari/product-heuristics-2020/issues/25
mentions that the drop down menu for the label brush shape isn't needed, because the circle is only shape needed.  I have removed it, and tried to remove anything that would include the square shape.

I am unsure if I removed too much or too little, so please look very carefully when reviewing!  Thanks!

### CODE REVIEW

**Comment by ppwadhwa** [comment]
and it looks like I need to address some of the changes I made to the tests.  Will get to that...

**Comment by sofroniewn** [comment]
Hi @ppwadhwa, thanks for getting this going. I definitely sympathize with the product heuristic assessment that having less user choice is good, and having less diverging code paths and attributes in napari is good too, so if we can make sure we don't loose anything critical in this change them I'm all for it. The square came first because it was easiest, but if we had started with a circle we might never have added the square!

Before merge there are probably four things I want to consider

1) Performance of the circle brush vs the square brush for a regular size image. The circle brush has a more complex code path that relies on numpy fancy indexing and so I thought would be slower. We did some benchmarking in the original PR that added the circle and it appeared to be faster than the square! See comment here https://github.com/napari/napari/pull/1429#issuecomment-658461298, possibly because we're editing only ~75% of the circle. We might just want to reconfirm this, or make sure we've tested under the full range of relevant conditions.

2) Painting in 3D, nD. There have been some discussions about how to paint in non-displayed dimensions, we could just broadcast the circle, or we could try and do an ellipsoid. I don't think this is any worse with circle vs square, we just might want to consider it.

3) Painting into none numpy arrays - like xarray or sparse arrays. Because the circle uses numpy fancy indexing we just want to make sure that we can always paint into these arrays too (the square is simpler, so should be less of a problem). We recently added support for xarray, and I think @DragaDoncila has now gotten sparse working too with circle, but we might want to double check.

4) Just curious if there are people out in the community who use square and prefer it for reasons we havn't anticipated, we might want to ask in zulip or at the next community meetings, or ask the alfa-cohort segmentation devs @neuromusic before dropping.

----------

Curious what @jni thinks about all of this?

We might also want to split this into two PRs too. One where we just drop the combobox from the UI and add a deprecation warning to the brush shape attribute, but leave all the code in place. We could merge that first and then two releases from now drop the attribute and all the code. That way we avoid any immediate breaking code changes. How does that sound?

**Comment by ppwadhwa** [comment]
@sofroniewn breaking it into 2 PRs sounds good. I will make some updates to just remove the dropbox and then start addressing the other concerns.  Thanks for the information, it was really helpful!

**Comment by DragaDoncila** [comment]
@ppwadhwa @sofroniewn I can confirm that the circle brush works just fine with sparse DOK arrays, so no objections here to removing the square brush.

**Comment by sofroniewn** [comment]
Just as an aside - this is the type of decision that if we had full telemetry on napari would be so much easier to make! There has been some discussion of this in #621, but imagine if we new the frequency of people using the square brush to the circle brush! (Now just because something only gets used 1% of the time doesn't mean we should take it away, but still good to know!!)

**Comment by sofroniewn** [comment]
@ppwadhwa can you add a deprecation warning in the property getter for `brush_shape`, it could look like the one here https://github.com/napari/napari/blob/4bb4075ed8836723f7810f1beccd210b3aa58360/napari/layers/base/base.py#L729-L741 and can go for +3 release from where we are now - so `will be removed in version 0.4.9`

**Comment by ppwadhwa** [comment]
> Just as an aside - this is the type of decision that if we had full telemetry on napari would be so much easier to make! There has been some discussion of this in #621, but imagine if we new the frequency of people using the square brush to the circle brush! (Now just because something only gets used 1% of the time doesn't mean we should take it away, but still good to know!!)

That's very true!  Having all that info would be really helpful.

**Comment by ppwadhwa** [comment]
> @ppwadhwa can you add a deprecation warning in the property getter for `brush_shape`, it could look like the one here
> 
> https://github.com/napari/napari/blob/4bb4075ed8836723f7810f1beccd210b3aa58360/napari/layers/base/base.py#L729-L741
> 
> and can go for +3 release from where we are now - so `will be removed in version 0.4.9`

Thank you for this!  I was just going to ask about the deprecation warning.  :)

File: napari/layers/labels/labels.py
Comment (sofroniewn) [inline napari/layers/labels/labels.py]: I think I would show the warning always, as it will just be assumed to be circle, so we won't need the brush shape attribute at all

Comment (sofroniewn) [reply napari/layers/labels/labels.py]: Maybe also actually put the warning on the setter too so if someone tries to set they will get one too

Comment (ppwadhwa) [reply napari/layers/labels/labels.py]: Ok, I removed that if statement and added the warning to the setter.  How is the wording of the warning?

Comment (sofroniewn) [reply napari/layers/labels/labels.py]: Thanks! How about adding on

```
"Afterward, only the circle brush shape will be available, and the layer.brush_shape attribute will be removed"
```

That makes it clear we'll be taking away the attribute too

**Comment by sofroniewn** [comment]
Our tests are all failing on the warning. The warning will need to be explicitly excluded here so as not to raise an error https://github.com/napari/napari/blob/95c0c3d053a07d1624e83c110add2520d1c9c911/pyproject.toml#L77-L87

When we drop the attribute we can then fix all the tests too

**Comment by ppwadhwa** [comment]
@sofroniewn I made those couple of changes.  Fingers crossed everything passes!  I did check locally, but you never know...

And thank you for pointing me to where to exclude the warning!  very helpful.  :)

**Review by sofroniewn — APPROVED** [review]
Ok this looks good to me! Maybe @jni can weigh in before merge if he has a strong argument for saving the square brush, but based on the benchmarking and our ability to use circle for some of the more advanced array types I'd be fine to see it go for the benefit of simplicity :-)

File: napari/layers/labels/labels.py
Comment (Carreau) [inline napari/layers/labels/labels.py:509]: :+1: thanks for making long and explicit deprecation warnings with versions number.

For my own understanding: Any policies on FutureWarning vs (Pending)DeprecationWarning ?

**Review by Carreau — APPROVED** [review]
+1 need a rebase though because of conflicts.

File: napari/layers/labels/labels.py
Comment (ppwadhwa) [reply napari/layers/labels/labels.py:509]: @sofroniewn What are your thoughts on the warning?  I just used the future warning because that was in the example you pointed to, but I am not sure if its what should be used here.

also, @goanpeca will the translation work on the warning?  should I change this to be wrapped in `trans_()`?

**Comment by ppwadhwa** [comment]
> +1 need a rebase though because of conflicts.

@Carreau thanks for catching that.  I just did a rebase.  :)

File: napari/layers/labels/labels.py
Comment (goanpeca) [reply napari/layers/labels/labels.py:509]: Please do :)

Comment (sofroniewn) [reply napari/layers/labels/labels.py:509]: > For my own understanding: Any policies on FutureWarning vs (Pending)DeprecationWarning ?

I think we did this once a while back and maybe I'm just copying and pasting every time. I'd be happy to defer to an alternative if its the recommended best practice, I don't have strong opinions here. Maybe we open up a new issue to discuss though, and if we decide to change quickly change all of them.

Comment (jni) [reply napari/layers/labels/labels.py:509]: DeprecationWarnings are not visible by default:

https://docs.python.org/3/library/exceptions.html#DeprecationWarning

We want these warnings to be visible to users without having to enable Python development mode.

**Review by jni — APPROVED** [review]
Yup, happy with this! 😊

File: napari/layers/labels/labels.py
Comment (tlambert03) [reply napari/layers/labels/labels.py:509]: they can potentially be made visible in, for instance, IPython or a script, as long as you adjust `stacklevel` appropriately (if it puts the warning at the level of `__main__`  which makes sense if the user directly called something that resulted in the warning) ...  Just in case that puts it back on the table (to use the warning that feels semantically more appropriate)?

Comment (tlambert03) [reply napari/layers/labels/labels.py:509]: so... with a `warn(..., DeprecationWarning, stacklevel=2)`, if someone directly accessed `labels_layer.brush_shape` the warning _would_ be visible to the user without dev mode.

File: pyproject.toml
Comment (tlambert03) [inline pyproject.toml:87]: this is fine too, but if it's just one or two tests that are failing because of this warning, you might also consider using a `with pytest.warngs(FutureWarning):` context right before the call that creates the warning.  (That would also remind us to remove the ignore it when the warning goes away eventually)

**Review by tlambert03 — APPROVED** [review]

File: napari/layers/labels/labels.py
Comment (sofroniewn) [reply napari/layers/labels/labels.py:509]: Makes sense, lets do in a follow up though

**Comment by sofroniewn** [comment]
Ok, this looks good to me! I will merge now, we can address the DeprecationWarning in a follow up PR. Thanks all!

**Comment by DragaDoncila** [comment]
@ppwadhwa I've just pulled down master and the brush shape warning pops up regardless of the brush shape being a circle, and it also fires every time you try to paint. I think we should tweak this a bit to only fire off the warning if the user tries to set the brush shape as square, rather than every time the brush shape is set/accessed.

**Comment by sofroniewn** [comment]
My recommendation @ppwadhwa  to fix is for us to internally use `layer._brush_shape` wherever we currently use `layer.brush_shape` and bypass the deprecation warning until we have removed the attribute entirely.

**Comment by sofroniewn** [comment]
Thanks for catching @DragaDoncila!

**Comment by ppwadhwa** [comment]
@sofroniewn @DragaDoncila I just made a new PR to fix this issue.  Please take a look when you get a chance!

File: napari/layers/labels/labels.py
Comment (jni) [reply napari/layers/labels/labels.py:509]: @sofroniewn @tlambert03 the issue though is that many users will not be setting these things directly in `__main__` but in functions/methods deep in their libraries. @tlambert03 points out that pytest sets dev mode and so you'd see them there, but not everyone has perfect test coverage...

**Pull request merged**
